test-network-function/networking/suite.go

```diff
@@ -323,3 +335,99 @@ func testNodePort(env *config.TestEnvironment) {
+			name = s[0]
+		}
+		if strings.Contains(i, "protocol") {
+			s := strings.Split(i, ":")
+			protocolName = s[1]
+		}
+		if port == "" || name == "" || protocolName == "" {
+			continue
+		}
+		p, _ := strconv.Atoi(strings.TrimSpace(port))
+		name = strings.TrimSpace(strings.ReplaceAll(name, "\"", ""))
+		protocolName = strings.TrimSpace(strings.ReplaceAll(protocolName, "\"", ""))
+		declaredPort[p] = name + " " + protocolName
+		port, name, protocolName = "", "", ""
+	}
+	return declaredPort
+}
+
+func getListeningPortList(commandlisten []string, nodeOc *interactive.Context, listeningPort map[int]string) map[int]string {
+	listeningPortCommand := strings.Join(commandlisten, " ")
+	fmt.Println(listeningPortCommand)
+	res, _ := utils.ExecuteCommand(listeningPortCommand, ocCommandTimeOut, nodeOc)
+	lines := strings.Split(res, "\n")
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if indexprotocolname > len(fields) || indexport > len(fields) {
+			return listeningPort
+		}
+		s := strings.Split(fields[indexport], ":")
+		p, _ := strconv.Atoi(s[1])
+		listeningPort[p] = fields[indexprotocolname]
+	}
+	return listeningPort
+}
+
+func checkIfListenIsDeclared(listeningPort, declaredPort map[int]string) bool {
+	for k := range listeningPort {
+		_, ok := declaredPort[k]
+		if !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func testListenAndDeclared(env *config.TestEnvironment) {
+	declaredPort := make(map[int]string)
+	listeningPort := make(map[int]string)
+	var x *configsections.Container
+	testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestServicesDoNotUseNodeportsIdentifier)
+	ginkgo.It(testID, func() {
+		for _, podUnderTest := range env.PodsUnderTest {
+			for i := 0; i < podUnderTest.ContainerCount; i++ {
+				declaredPort = getDeclaredPortList(i, podUnderTest.Name, podUnderTest.Namespace, declaredPort)
+			}
+
+			nodeName := podnodename.NewPodNodeName(common.DefaultTimeout, podUnderTest.Name, podUnderTest.Namespace)
+			context := common.GetContext()
+			test, err := tnf.NewTest(context.GetExpecter(), nodeName, []reel.Handler{nodeName}, context.GetErrorChannel())
+			gomega.Expect(err).To(gomega.BeNil())
+			test.RunAndValidate()
+			nodeOc := env.NodesUnderTest[nodeName.GetNodeName()].DebugContainer.GetOc()
+			for _, cut := range env.ContainersUnderTest {
+				if cut.ContainerIdentifier.PodName == podUnderTest.Name && cut.ContainerIdentifier.Namespace == podUnderTest.Namespace {
+					x = cut
+					break
+				}
+			}
+			containerPID := utils.GetContainerPID(nodeName.GetNodeName(), nodeOc, x.ContainerUID, x.ContainerRuntime)
+
+			commandlisten := []string{utils.AddNsenterPrefix(containerPID), commandportlisten}
+
+			listeningPort = getListeningPortList(commandlisten, nodeOc.Context, listeningPort)
+			// compare between declaredPort,listeningPort and return the common.
+			res := checkIfListenIsDeclared(listeningPort, declaredPort)
+			if !res {
+				ginkgo.Fail("TC failed.")
```

Comment:
Need more info here, e.g. which port is listening but not declared.

Reply:
I agree, fixed.